stf/model/trainer/lightgbm/lightgbm.py

```diff
@@ -0,0 +1,452 @@
+    ):
+        """Objective function that picks the optimal training_duration value.
+
+        This function should be used after the hyper_params_objective has been completed
+        by optuna.
+
+        Args:
+            trial: optuna trial object that is passed on during hyper parameter
+                optimalisation.
+            error_function (callable): Function to calculate the error metric to be
+                optimized, preferably one from stf.model.metrics.
+            unprocessed_data (pandas.DataFrame): Data and features that have not yet
+                been pre-processed.
+            training_durations_days (list of int): Candidate training durations.
+            featuresets (dict): All feature sets: with keys the featureset_name and
+                values a list of feature names.
+
+        Returns:
+            float: Error metric value that is either minimized or maximized by the
+            optuna study.
+
+        """
+
+        # Make selection of training-duration
+        training_period = trial.suggest_categorical(
+            "training_period_days", training_durations_days
+        )
+
+        # Update hyper parameters
+        self.hyper_parameters.update(optimized_parameters)
+
+        # Set training duration and shrink incoming data accordingly
+        datetime_start = (
+            unprocessed_data.index.max() - timedelta(days=training_period)
+        ).replace(tzinfo=pytz.UTC)
+        shortened_data = unprocessed_data.loc[unprocessed_data.index > datetime_start]
+
+        # Pre-process data
+        clean_shortened_data_with_all_features = pre_process_data(shortened_data)
+
+        # Apply optimized featureset
+        featureset_name = optimized_parameters["featureset_name"]
+
+        featureset = featuresets[featureset_name]
+        total_data = remove_features_not_in_set(
+            clean_shortened_data_with_all_features, featureset=featureset
+        )
+
+        # Split data in train, test and validation sets, note we are using the
+        # backtest option here because we assume hyperparameters do not change
+        # much over time
+        train_data, validation_data, test_data = split_data_train_validation_test(
+            total_data, test_fraction=0.1, validation_fraction=0.1, back_test=True,
+        )
+
+        # Train model
+        model = self.train(train_data, validation_data)
+
+        # Make prediction on test data and prepare dataframes for comparison
+        prediction = pd.DataFrame(
+            model.predict(
+                lgb.Dataset(test_data.iloc[:, 1:].drop("Horizon", axis=1, errors="ignore")), num_iteration=model.best_iteration
+            )
+        )
+        realised = pd.DataFrame(test_data.iloc[:, 0])
+        prediction.index = realised.index  # Set correct DateTime index
+
+        # Calculate error metric
+        error = error_function(realised.iloc[:, 0], prediction.iloc[:, 0])
+        return error
+
+    def hyper_params_objective(
```

Comment:
This objective function seems to be xgb specific, please consider if you want to keep it her or want to adapt to lightgbm

Reply:
it actually is already updates since the error function for lightgbm is slightly different compared to xgb